### ISSUE TITLE
Fix for file uploads

### ISSUE THREAD

**Issue by francisjervis** [milesrichardson/ParsePy#157]
This was fixed in the comments for #15 by @treeguard but no PR was pushed to this branch:
115 data = data.encode('utf-8') > data = data
and
124 request = Request(url, data, headers) > request = Request(url.encode('utf-8'), data, headers)
Thanks!

**Comment by milesrichardson** [comment]
@francisjervis Can you confirm that this fix works on your setup? If so, would you mind submitting a PR? I don't have a self hosted parse-server with file support setup right now, so reproducing/testing would take longer than a few seconds.

If you need help with the workflow of editing the repository in your development environment and submitting the PR, what I recommend is: forking the repository, cloning it to the current directory, then `sys.path.append(currentDirectory + 'ParsePy')` before importing. Then you can edit the files in the `ParsePy` until you get it to work, then commit changes, `git push origin master` to your fork, and finally submit a PR on github.

**Comment by milesrichardson** [comment]
@francisjervis I just attempted that fix and it did not work. File support has been broken in this library for a long time...

Way back before I inherited this library, when we were still using parse-server, I had some workaround code for uploading files to the rest API:

```

PARSE_HEADERS = {
    'X-Parse-Application-Id': 'xxxxx',
    'X-Parse-Master-Key': 'yyyyyy',
    'Content-Type': 'application/json'
}


def upload_media(rawdata, parsefn, retry=0):
    ''' Upload a media file to parse, and return value suitable for object.

        upload_media will infer the media type (mp4, png, or jpg) from the
        file extension of parsefn.
    '''
    headers = PARSE_HEADERS
    if 'mp4' in parsefn:
        headers['Content-Type'] = 'video/mp4'
    elif 'png' in parsefn:
        headers['Content-Type'] = 'image/png'
    else:
        headers['Content-Type'] = 'image/jpeg'

    conn = httplib.HTTPSConnection('api.parse.com', 443)
    conn.connect()
    conn.request('POST', '/1/files/' + parsefn, rawdata, headers)

    rval = False
    try:
        media_name = json.loads(conn.getresponse().read())['name']
        rval = media_name
    except:
        if retry >= 5:
            rval = False
        else:
            rval = upload_media(rawdata, parsefn, retry + 1)

    conn.close()
    return rval

def upload_to_parse(parse_object, class_name, retry=0):
    ''' Upload a raw object to parse, in dictionary format as expected.

    '''
    conn = httplib.HTTPSConnection('api.parse.com', 443)
    conn.connect()
    conn.request('POST', '/1/classes/%s' % class_name,
                 json.dumps(parse_object),
                 PARSE_HEADERS)

    try:
        result = json.loads(conn.getresponse().read())
        if 'error' in result.keys():
            print 'FAILED upload: %s' % str(result)
            raise Exception
        else:
            rval = True
    except:
        if retry >= 5:
            rval = False
        else:
            rval = upload_to_parse(parse_object, class_name, retry+1)

    conn.close()

    return rval
```

Feel free to adapt that to your use case. Or, even better, integrate it into the library and submit a PR. :)

**Comment by milesrichardson** [comment]
@francisjervis Fixed (safely, using a check to see if class is a File). :) 

pull from master to use

**Comment by newdev7** [comment]
Hi,

I'm using Python 3.6.0 for my project. I'm using this library (version='0.2.20170114') for storing the details in parse. When I tried to upload the file in Parse using File, it returns error saying 
`<urlopen error unknown url type: b'https>`.

The issue is with the url in the request. In the connection.py file, line 128, you are converting the url string to bytes using `url.encode('utf-8')`. This creates the issue. If I remove the encode and send the url as it is, then there is no issue. So I changed the code from `request = Request(url.encode('utf-8'), data, headers)` to `request = Request(url, data, headers)`. Now it is working.
